dockerunit-core/src/main/java/com/github/qzagarese/dockerunit/internal/lifecycle/DockerUnitBefore.java

```diff
@@ -19,21 +26,62 @@
 	private final Statement next;
 	private final DiscoveryProvider discoveryProvider;
 	private final ServiceContextBuilder contextBuilder;
-	private final DependencyDescriptor descriptor;
+	private final UsageDescriptor descriptor;
 	
 	@Override
 	public void evaluate() throws Throwable {
-        ServiceContext context = contextBuilder.buildContext(descriptor);
-        runner.setContext(method, context);
-        if(!context.allHealthy()) {
-        	throw new RuntimeException(context.getFormattedErrors());
-        }
-        context = discoveryProvider.populateRegistry(context.merge(runner.getClassContext()));
-        runner.setContext(method, context);
-        if(!context.allHealthy()) {
-        	throw new RuntimeException(context.getFormattedErrors());
+	    
+	    List<ServiceContext> serviceContexts = new ArrayList<>();
+	    
+	    descriptor.getDependencies().forEach(svc -> {
+	        // Build containers
+	        ServiceContext singleSvcContext = contextBuilder.buildServiceContext(svc);
+	        if(!singleSvcContext.allHealthy()) {
+	            throw new RuntimeException(singleSvcContext.getFormattedErrors());
+	        }
+	        
+	        // Merge with class context in case there are more instances of this service at class level
+	        singleSvcContext = singleSvcContext.merge(findEquivalentInClass(svc.getNamed().value(), runner.getClassContext()));
+	        
+	        // Perform discovery
+	        singleSvcContext = discoveryProvider.populateRegistry(singleSvcContext);
+	        if(!singleSvcContext.allHealthy()) {
+                throw new RuntimeException(singleSvcContext.getFormattedErrors());
+            }
```

Comment:
Every time `populateRegistry` is called this check (`ServiceContext::allHealthy` is done). Would it make sense to do that as part of the `populateRegistry` and have it throw the exception?

Same applies to `buildServiceContext`, in fact why check twice for the health status? (after building the ctx and after populating the registry)

Reply:
I see your point, but I prefer having the check in here as it helps me keep track of the main phases:

1. Create container -> check health
2. Perform discovery -> check health

I can extract the if in private method if you think that it is more readable.